app/models/vacancy.rb

```diff
@@ -187,16 +187,16 @@ def set_postcode_from_mean_geolocation!
     # version of that geolocation (i.e. a postcode).
     return if central_office?
 
-    coords = organisations.reload.schools.filter_map(&:geolocation).map { |loc| [loc.x, loc.y] }.reject { |coord| coord == [0, 0] }
-    return unless coords.any?
-
-    mean_geolocation = [(coords.sum(&:first) / coords.length), (coords.sum(&:second) / coords.length)]
-    self.postcode_from_mean_geolocation = if at_one_school?
-                                            reload
-                                            organisation.postcode
-                                          else
-                                            Geocoding.new(mean_geolocation.to_a).postcode_from_coordinates || postcode_from_mean_geolocation
-                                          end
+    if at_one_school?
+      self.postcode_from_mean_geolocation = organisation.postcode
+    else
+      schools = organisations.schools.where.not(geopoint: nil)
+      centroid = schools.pluck(Arel.sql("ST_Centroid(ST_Collect(geopoint::geometry))::geography")).first
+
+      return unless centroid
+
+      self.postcode_from_mean_geolocation = Geocoding.new([centroid.x, centroid.y]).postcode_from_coordinates || postcode_from_mean_geolocation
```

Comment:
Do we want to not override any existing value for this field? If so you can write this more concisely as:

```ruby
self.postcode_from_mean_geolocation = Geocoding.new([centroid.x, centroid.y]).postcode_from_coordinates
```

Reply:
This is to avoid overwriting the attribute with nil when the Geocoder returns a nil response.

Reply:
I'm not sure how your suggestion works?

Reply:
Apologies, that was pre-coffee! I meant:

```ruby
self.postcode_from_mean_geolocation ||= Geocoding.new([centroid.x, centroid.y]).postcode_from_coordinates
```

(with the `||=` operator)

Reply:
Sorry that was pre-coffee and I typed it wrong, and also I see what you mean now.

I think I'd prefer a slightly longer but more explicit way of writing this bit, so it's more obvious what the intention is, e.g.:

```ruby
postcode = Geocoding.new([centroid.x, centroid.y]).postcode_from_coordinates
return unless postcode

self.postcode_from_mean_geolocation = postcode
```